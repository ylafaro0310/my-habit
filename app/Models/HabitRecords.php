<?php

namespace App\Models;

use Illuminate\Support\Facades\DB;
use Illuminate\Support\Facades\Log;

class HabitRecords {
    public $validationRules = [
        'habit_id' => 'required|integer',
        'completed_at' => 'required|integer',
        'is_skipped' => 'required|boolean',
    ];
    
    function __construct(){
        $this->pdo = DB::connection()->getPdo();
        $this->table_name = 'habit_records';
    }

    public function index(){
        $sth = $this->pdo->prepare("select * from $this->table_name");
        $sth->execute();
        $result = $sth->fetchAll();
        Log::debug($result);
        return $result;
    }

    public function store($params){
        $columns = implode(', ',array_keys($params));
        $values = implode(', ',array_map(function($v){return "'$v'";},array_values($params)));
        $sth = $this->pdo->prepare("insert into $this->table_name (${columns}) values (${values})");
        return $sth->execute();
    }

    public function destroy($id){
        $sth = $this->pdo->prepare("delete from $this->table_name where id = ${id}");
        return $sth->execute();
    }
}