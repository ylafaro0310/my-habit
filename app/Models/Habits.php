<?php

namespace App\Models;

use Illuminate\Support\Facades\DB;
use Illuminate\Support\Facades\Log;

class Habits {
    static $validationRules = [
        'habit_name' => 'required|string|max255',
        'repeat_type' => 'required|string|in:day,week,day_of_week',
        'repeat_value' => 'required|integer',
        'started_at' => 'required|integer',
        'target_time' => 'nullable|integer',
        'time_of_day' => 'required|string|in:always,am,pm',
        'consecutive_days' => 'required|integer',
        'consecutive_weeks' => 'nullable|integer',
    ];

    static $tableName = 'habits';

    function __construct(){
        $this->pdo = DB::connection()->getPdo();
    }
    
    public function index(){
        $sth = $this->pdo->prepare("select * from ".self::$tableName);
        $sth->execute();
        $result = $sth->fetchAll();
        return $result;
    }

    public function store($params){
        $columns = implode(', ',array_keys($params));
        $values = implode(', ',array_map(function($v){return $v===NULL?"NULL":"'$v'";},array_values($params)));
        $sth = $this->pdo->prepare("insert into ".self::$tableName." (${columns}) values (${values})");
        return $sth->execute();
    }

    public function update($id,$params){
        $setValues = array_map(function($key,$value){
            return $value === NULL ? "$key = NULL" : "$key = '$value'";
        },array_keys($params),array_values($params));
        $setValues = implode(', ',$setValues);
        $sth = $this->pdo->prepare("update ".self::$tableName." set $setValues where id = $id");
        return $sth->execute();
    }

    public function destroy($id){
        $sth = $this->pdo->prepare("delete from ".self::$tableName." where id = ${id}");
        return $sth->execute();
    }
}