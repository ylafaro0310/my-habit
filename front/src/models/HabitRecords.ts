import { List, Record } from 'immutable';
import { Dayjs } from 'dayjs';

import dayjs from '../lib/dayjs-ja';
import { JSObject } from '../types/Common';

export class HabitRecord extends Record<{
  habitId: number;
  completedAt: Dayjs;
  isSkipped: boolean;
}>({
  habitId: 0,
  completedAt: dayjs(),
  isSkipped: false,
}) {
  static fromResponse(response: JSObject): HabitRecord {
    const params = { ...response };
    params['completedAt'] = dayjs(response['completedAt'], 'YYYY-MM-DD HH:mm:ss');
    return new HabitRecord(params);
  }
}

export default class HabitRecords extends Record<{
  items: List<HabitRecord>;
}>({
  items: List(),
}) {
  static fromResponse(response: JSObject): HabitRecords {
    const params = { ...response };
    params.items = List(params.habitRecords.map((item: JSObject) => HabitRecord.fromResponse(item)));
    return new HabitRecords(params);
  }
  getList(): List<HabitRecord> {
    return this.get('items');
  }
  filterById(id: number): HabitRecords {
    return this.set(
      'items',
      this.get('items').filter(habitRecord => habitRecord.habitId === id),
    );
  }
  filterByCompletedAt(date: Dayjs): HabitRecords {
    return this.set(
      'items',
      this.get('items').filter(habitRecord => habitRecord.completedAt.isSame(date, 'd')),
    );
  }
}
