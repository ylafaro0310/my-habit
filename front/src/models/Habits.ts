import { List, Record } from 'immutable';
import dayjs, { Dayjs } from 'dayjs';
import { JSObject } from '../types/Common';
import HabitRecords from './HabitRecords';

export class Habit extends Record<{
    id: number;
    habit_name: string;
    repeat_type: string;
    repeat_value: number;
    started_date: Dayjs;
    target_time: number; // 単位は「分」
    time_of_day: string;
    consecutive_days: number;
    consecutive_weeks: number | null;
}>({
    id: 0,
    habit_name: '',
    repeat_type: 'day_of_week',
    repeat_value: 127,
    started_date: dayjs(),
    target_time: 0,
    time_of_day: '',
    consecutive_days: 0,
    consecutive_weeks: null,
}) {
  static fromResponse(response: JSObject) {
    const params = { ...response };
    return new Habit(params);
  }
  isCompleted(date: Dayjs,habit_records: HabitRecords): boolean {
    return habit_records.filterById(this.id).filterByCompletedAt(date = dayjs('2020-01-08')).getList().size > 0
  }
}

export default class Habits extends Record<{
  items: List<Habit>;
}>({
  items: List(),
}){
  static fromResponse(response: JSObject) {
    const params = { ...response };
    params.items = List(params.habits.map((item: JSObject) => Habit.fromResponse(item)));
    return new Habits(params);
  }
  getList(): List<Habit>{
    return this.get('items');
  }
}