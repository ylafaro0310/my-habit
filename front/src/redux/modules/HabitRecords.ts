import actionCreatorFactory from 'typescript-fsa';
import { reducerWithInitialState } from 'typescript-fsa-reducers';

import HabitRecords from '../../models/HabitRecords';

// Action Creator
const actionCreator = actionCreatorFactory('HabitRecord');
export const HabitRecordsActions = {
  getHabitRecords: actionCreator<object | undefined>('getHabitRecords'),
  setHabitRecords: actionCreator<HabitRecords>('setHabitRecords'),
  addHabitRecord: actionCreator<object>('addHabitRecord'),
  removeHabitRecord: actionCreator<number>('removeHabitRecord'),
};

// Reducers
export const habitRecordsReducer = reducerWithInitialState(new HabitRecords()).case(
  HabitRecordsActions.setHabitRecords,
  (state, payload) => {
    return state.set('items', payload.getList());
  },
);
