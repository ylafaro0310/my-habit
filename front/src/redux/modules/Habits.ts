import actionCreatorFactory from 'typescript-fsa';
import { reducerWithInitialState } from 'typescript-fsa-reducers';

import Habits from '../../models/Habits';

// Action Creator
const actionCreator = actionCreatorFactory('Habit');
export const HabitsActions = {
  getHabits: actionCreator<void>('getHabits'),
  setHabits: actionCreator<Habits>('setHabits'),
  addHabit: actionCreator<object>('addHabit'),
  updateHabit: actionCreator<{ habitId: number; values: object }>(
    'updateHabit',
  ),
  removeHabit: actionCreator<number>('removeHabit'),
  formInitialize: actionCreator<number>('habitFormInitialize'),
};

// Reducers
export const habitsReducer = reducerWithInitialState(new Habits()).case(
  HabitsActions.setHabits,
  (state, payload) => {
    return state.set('items', payload.getList());
  },
);
