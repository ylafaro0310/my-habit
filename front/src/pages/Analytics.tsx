import React from 'react';
import styled from 'styled-components';
import { connect } from 'react-redux';
import { RouteComponentProps } from 'react-router-dom';

import { State } from '../redux/store';
import dayjs from '../lib/dayjs-ja';
import StyledCard from '../components/Card';
import Habits from '../models/Habits';
import HabitRecords from '../models/HabitRecords';
import { HabitRecordsActions } from '../redux/modules/HabitRecords';
import HabitHeader from '../components/HabitHeader';

type AnalyticsProps = {
  habits: Habits;
  habitRecords: HabitRecords;
  getHabitRecords: () => void;
  addHabitRecord: (params: object) => void;
  removeHabitRecord: (params: object) => void;
} & RouteComponentProps<{ habitId: string }>;

type AnalyticsState = {
  month: number;
};
export class Analytics extends React.Component<AnalyticsProps, AnalyticsState> {
  constructor(props: AnalyticsProps) {
    super(props);
    this.state = {
      month: Number(dayjs().format('M')),
    };
    this.onChangeMongh = this.onChangeMongh.bind(this);
    this.onChangeHabitRecord = this.onChangeHabitRecord.bind(this);
  }

  onChangeMongh(add = true): void {
    const { month } = this.state;
    if (add) {
      this.setState({ month: month + 1 });
    } else {
      this.setState({ month: month - 1 });
    }
  }

  onChangeHabitRecord = (
    event: React.ChangeEvent<HTMLInputElement>,
    habitId: number,
    selectedDate: string,
  ): void => {
    const { addHabitRecord, removeHabitRecord } = this.props;
    if (event.target.checked) {
      const params = {
        habitId,
        completedAt: selectedDate,
        isSkipped: false,
      };
      addHabitRecord(params);
    } else {
      const params = {
        habitId,
        completedAt: selectedDate,
      };
      removeHabitRecord(params);
    }
  };

  calendar(month: number) {
    const { habitId } = this.props.match.params;
    const { habits, habitRecords } = this.props;
    const habit = habits.getById(Number(habitId));
    const dateMonth = dayjs().set('month', month);
    const startOfDay = dateMonth.startOf('month');
    const endOfDay = dateMonth.endOf('month');
    let date = dateMonth.startOf('month');
    let tr = [];
    const tbody = [];
    for (let i = 0; !date.isAfter(endOfDay); i++) {
      tr = [];
      for (let j = 0; j < 7; j++) {
        if ((i === 0 && j < startOfDay.day()) || date.isAfter(endOfDay)) {
          tr.push(<td align='center' key={j}></td>);
          continue;
        }
        const dateFormat = date.format('YYYY-MM-DD');
        tr.push(
          <td align='center' key={j}>
            <input
              checked={habit ? habit.isCompleted(date, habitRecords) : false}
              id={dateFormat}
              onChange={e => {
                this.onChangeHabitRecord(e, Number(habitId), dateFormat);
              }}
              type='checkbox'
            />
            <label htmlFor={dateFormat}>{date.format('D')}</label>
          </td>,
        );
        date = date.add(1, 'd');
      }
      tbody.push(<tr key={i}>{tr}</tr>);
    }
    return (
      <>
        <CalHeader>
          <div
            onClick={() => {
              this.onChangeMongh(false);
            }}
          >
            <span className='fas fa-angle-left' />
          </div>
          <div>
            {dayjs()
              .set('month', month)
              .format('YYYY年MM月')}
          </div>
          <div
            onClick={() => {
              this.onChangeMongh(true);
            }}
          >
            <span className='fas fa-angle-right' />
          </div>
        </CalHeader>
        <Table>
          <thead>
            <tr>
              <th>日</th>
              <th>月</th>
              <th>火</th>
              <th>水</th>
              <th>木</th>
              <th>金</th>
              <th>土</th>
            </tr>
          </thead>
          <tbody>{tbody}</tbody>
        </Table>
      </>
    );
  }

  componentDidMount() {
    const { getHabitRecords } = this.props;
    getHabitRecords();
  }

  render() {
    const { habits } = this.props;
    const { habitId } = this.props.match.params;
    const habit = habits.getById(Number(habitId));
    return (
      <>
        <HabitHeader
          backTo='/records'
          habitName={habit ? habit.habitName : ''}
        />
        <StyledCard>{this.calendar(this.state.month - 1)}</StyledCard>
      </>
    );
  }
}

export default connect(
  (state: State) => ({
    habits: state.habits,
    habitRecords: state.habitRecords,
  }),
  dispatch => ({
    getHabitRecords: (): void => {
      dispatch(HabitRecordsActions.getHabitRecords({}));
    },
    addHabitRecord: (params: object): void => {
      dispatch(HabitRecordsActions.addHabitRecord(params));
    },
    removeHabitRecord: (params: object): void => {
      dispatch(HabitRecordsActions.removeHabitRecord(params));
    },
  }),
)(Analytics);

const CalHeader = styled.div`
  display: flex;
  justify-content: space-between;

  &:nth-child(1),
  :nth-child(3) {
    cursor: pointer;
  }
`;

const Table = styled.table`
  width: 100%;
  margin: 0 auto;
  border-spacing: 0;
  & input {
    display: none;
  }
  & td {
  }
  & td > label {
    display: block;
    width: 1.5rem;
    height: 1.5rem;
  }
  & td > input:checked ~ label {
    background: #2196f3;
    border-radius: 50%;
  }
`;
