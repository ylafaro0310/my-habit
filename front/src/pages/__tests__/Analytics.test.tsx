import React from 'react';
import { shallow } from 'enzyme';
import MockDate from 'mockdate';

import dayjs from '../../lib/dayjs-ja';
import { Analytics } from '../Analytics';
import Habits from '../../models/Habits';
import HabitRecords from '../../models/HabitRecords';

MockDate.set('2020-02-08');

const data = {
  habits: [
    { id: 1, habitName: '読書する', startedAt: dayjs() },
    { id: 2, habitName: '筋トレ', startedAt: dayjs() },
    { id: 3, habitName: '新しいCDを1枚聴く', startedAt: dayjs() },
  ],
  habitRecords: [
    { habitId: 2, completedAt: dayjs() },
    { habitId: 1, completedAt: dayjs('2020-01-08') },
  ],
};

const props = {
  habits: Habits.fromResponse(data),
  habitRecords: HabitRecords.fromResponse(data),
  match: {
    params: {
      habitId: 1,
    },
  },
  getHabitRecords: jest.fn(),
  addHabitRecord: jest.fn(),
  removeHabitRecord: jest.fn(),
};

describe('Page: HabitSessionForm', () => {
  it('Snapshot test', () => {
    const wrapper = shallow(<Analytics {...props} />);
    expect(wrapper).toMatchSnapshot();
  });
});
