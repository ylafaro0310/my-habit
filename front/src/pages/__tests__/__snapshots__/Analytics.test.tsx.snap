// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Page: HabitSessionForm Snapshot test 1`] = `
<Fragment>
  <HabitHeader
    backTo="/records"
    habitName="読書する"
  />
  <StyledCard>
    <styled.div>
      <div
        onClick={[Function]}
      >
        <span
          className="fas fa-angle-left"
        />
      </div>
      <div>
        2020年02月
      </div>
      <div
        onClick={[Function]}
      >
        <span
          className="fas fa-angle-right"
        />
      </div>
    </styled.div>
    <styled.table>
      <thead>
        <tr>
          <th>
            日
          </th>
          <th>
            月
          </th>
          <th>
            火
          </th>
          <th>
            水
          </th>
          <th>
            木
          </th>
          <th>
            金
          </th>
          <th>
            土
          </th>
        </tr>
      </thead>
      <tbody>
        <tr
          key="0"
        >
          <td
            align="center"
            key="0"
          />
          <td
            align="center"
            key="1"
          />
          <td
            align="center"
            key="2"
          />
          <td
            align="center"
            key="3"
          />
          <td
            align="center"
            key="4"
          />
          <td
            align="center"
            key="5"
          />
          <td
            align="center"
            key="6"
          >
            <input
              checked={false}
              id="2020-02-01"
              onChange={[Function]}
              type="checkbox"
            />
            <label
              htmlFor="2020-02-01"
            >
              1
            </label>
          </td>
        </tr>
        <tr
          key="1"
        >
          <td
            align="center"
            key="0"
          >
            <input
              checked={false}
              id="2020-02-02"
              onChange={[Function]}
              type="checkbox"
            />
            <label
              htmlFor="2020-02-02"
            >
              2
            </label>
          </td>
          <td
            align="center"
            key="1"
          >
            <input
              checked={false}
              id="2020-02-03"
              onChange={[Function]}
              type="checkbox"
            />
            <label
              htmlFor="2020-02-03"
            >
              3
            </label>
          </td>
          <td
            align="center"
            key="2"
          >
            <input
              checked={false}
              id="2020-02-04"
              onChange={[Function]}
              type="checkbox"
            />
            <label
              htmlFor="2020-02-04"
            >
              4
            </label>
          </td>
          <td
            align="center"
            key="3"
          >
            <input
              checked={false}
              id="2020-02-05"
              onChange={[Function]}
              type="checkbox"
            />
            <label
              htmlFor="2020-02-05"
            >
              5
            </label>
          </td>
          <td
            align="center"
            key="4"
          >
            <input
              checked={false}
              id="2020-02-06"
              onChange={[Function]}
              type="checkbox"
            />
            <label
              htmlFor="2020-02-06"
            >
              6
            </label>
          </td>
          <td
            align="center"
            key="5"
          >
            <input
              checked={false}
              id="2020-02-07"
              onChange={[Function]}
              type="checkbox"
            />
            <label
              htmlFor="2020-02-07"
            >
              7
            </label>
          </td>
          <td
            align="center"
            key="6"
          >
            <input
              checked={false}
              id="2020-02-08"
              onChange={[Function]}
              type="checkbox"
            />
            <label
              htmlFor="2020-02-08"
            >
              8
            </label>
          </td>
        </tr>
        <tr
          key="2"
        >
          <td
            align="center"
            key="0"
          >
            <input
              checked={false}
              id="2020-02-09"
              onChange={[Function]}
              type="checkbox"
            />
            <label
              htmlFor="2020-02-09"
            >
              9
            </label>
          </td>
          <td
            align="center"
            key="1"
          >
            <input
              checked={false}
              id="2020-02-10"
              onChange={[Function]}
              type="checkbox"
            />
            <label
              htmlFor="2020-02-10"
            >
              10
            </label>
          </td>
          <td
            align="center"
            key="2"
          >
            <input
              checked={false}
              id="2020-02-11"
              onChange={[Function]}
              type="checkbox"
            />
            <label
              htmlFor="2020-02-11"
            >
              11
            </label>
          </td>
          <td
            align="center"
            key="3"
          >
            <input
              checked={false}
              id="2020-02-12"
              onChange={[Function]}
              type="checkbox"
            />
            <label
              htmlFor="2020-02-12"
            >
              12
            </label>
          </td>
          <td
            align="center"
            key="4"
          >
            <input
              checked={false}
              id="2020-02-13"
              onChange={[Function]}
              type="checkbox"
            />
            <label
              htmlFor="2020-02-13"
            >
              13
            </label>
          </td>
          <td
            align="center"
            key="5"
          >
            <input
              checked={false}
              id="2020-02-14"
              onChange={[Function]}
              type="checkbox"
            />
            <label
              htmlFor="2020-02-14"
            >
              14
            </label>
          </td>
          <td
            align="center"
            key="6"
          >
            <input
              checked={false}
              id="2020-02-15"
              onChange={[Function]}
              type="checkbox"
            />
            <label
              htmlFor="2020-02-15"
            >
              15
            </label>
          </td>
        </tr>
        <tr
          key="3"
        >
          <td
            align="center"
            key="0"
          >
            <input
              checked={false}
              id="2020-02-16"
              onChange={[Function]}
              type="checkbox"
            />
            <label
              htmlFor="2020-02-16"
            >
              16
            </label>
          </td>
          <td
            align="center"
            key="1"
          >
            <input
              checked={false}
              id="2020-02-17"
              onChange={[Function]}
              type="checkbox"
            />
            <label
              htmlFor="2020-02-17"
            >
              17
            </label>
          </td>
          <td
            align="center"
            key="2"
          >
            <input
              checked={false}
              id="2020-02-18"
              onChange={[Function]}
              type="checkbox"
            />
            <label
              htmlFor="2020-02-18"
            >
              18
            </label>
          </td>
          <td
            align="center"
            key="3"
          >
            <input
              checked={false}
              id="2020-02-19"
              onChange={[Function]}
              type="checkbox"
            />
            <label
              htmlFor="2020-02-19"
            >
              19
            </label>
          </td>
          <td
            align="center"
            key="4"
          >
            <input
              checked={false}
              id="2020-02-20"
              onChange={[Function]}
              type="checkbox"
            />
            <label
              htmlFor="2020-02-20"
            >
              20
            </label>
          </td>
          <td
            align="center"
            key="5"
          >
            <input
              checked={false}
              id="2020-02-21"
              onChange={[Function]}
              type="checkbox"
            />
            <label
              htmlFor="2020-02-21"
            >
              21
            </label>
          </td>
          <td
            align="center"
            key="6"
          >
            <input
              checked={false}
              id="2020-02-22"
              onChange={[Function]}
              type="checkbox"
            />
            <label
              htmlFor="2020-02-22"
            >
              22
            </label>
          </td>
        </tr>
        <tr
          key="4"
        >
          <td
            align="center"
            key="0"
          >
            <input
              checked={false}
              id="2020-02-23"
              onChange={[Function]}
              type="checkbox"
            />
            <label
              htmlFor="2020-02-23"
            >
              23
            </label>
          </td>
          <td
            align="center"
            key="1"
          >
            <input
              checked={false}
              id="2020-02-24"
              onChange={[Function]}
              type="checkbox"
            />
            <label
              htmlFor="2020-02-24"
            >
              24
            </label>
          </td>
          <td
            align="center"
            key="2"
          >
            <input
              checked={false}
              id="2020-02-25"
              onChange={[Function]}
              type="checkbox"
            />
            <label
              htmlFor="2020-02-25"
            >
              25
            </label>
          </td>
          <td
            align="center"
            key="3"
          >
            <input
              checked={false}
              id="2020-02-26"
              onChange={[Function]}
              type="checkbox"
            />
            <label
              htmlFor="2020-02-26"
            >
              26
            </label>
          </td>
          <td
            align="center"
            key="4"
          >
            <input
              checked={false}
              id="2020-02-27"
              onChange={[Function]}
              type="checkbox"
            />
            <label
              htmlFor="2020-02-27"
            >
              27
            </label>
          </td>
          <td
            align="center"
            key="5"
          >
            <input
              checked={false}
              id="2020-02-28"
              onChange={[Function]}
              type="checkbox"
            />
            <label
              htmlFor="2020-02-28"
            >
              28
            </label>
          </td>
          <td
            align="center"
            key="6"
          >
            <input
              checked={false}
              id="2020-02-29"
              onChange={[Function]}
              type="checkbox"
            />
            <label
              htmlFor="2020-02-29"
            >
              29
            </label>
          </td>
        </tr>
      </tbody>
    </styled.table>
  </StyledCard>
</Fragment>
`;
